cmd/webhook/main.go

```diff
@@ -37,11 +42,45 @@ func main() {
 	webhook.SetupInClusterClient()
 
 	/* register handlers */
+	var httpServer *http.Server
 	http.HandleFunc("/mutate", webhook.MutateHandler)
 
-	/* start serving */
-	err := http.ListenAndServeTLS(fmt.Sprintf("%s:%d", *address, *port), *cert, *key, nil)
-	if err != nil {
-		glog.Fatalf("error starting web server: %s", err)
+	/* start webhook server */
+	go func() {
+		for {
+			httpServer = &http.Server{
+				Addr: fmt.Sprintf("%s:%d", *address, *port),
+			}
+			err := httpServer.ListenAndServeTLS(*cert, *key)
+			if err != nil {
+				if err == http.ErrServerClosed {
+					glog.Info("restarting server")
+					continue
+				} else {
+					glog.Fatalf("error starting web server: %s", err.Error())
+					break
+				}
+			}
+		}
+	}()
+
+	/* watch the cert file and restart http sever if the file updated. */
+	oldHashVal := ""
+	for {
+		hasher := sha512.New()
+		s, err := ioutil.ReadFile(*cert)
+		hasher.Write(s)
+		if err != nil {
+			glog.Fatalf("failed to read file %s: %s", *cert, err)
+		}
+		newHashVal := hex.EncodeToString(hasher.Sum(nil))
+		if oldHashVal != "" && newHashVal != oldHashVal {
+			glog.Info("get cert file update")
+			if err := httpServer.Shutdown(context.Background()); err != nil {
+				glog.Fatalf("http server shutdown: %v", err)
+			}
+		}
+		oldHashVal = newHashVal
+		time.Sleep(1 * time.Second)
```

Comment:
do you think that we could watch for the changes in the filesystem (inotify) or watch Kubernetes secret resource instead of recalculating hash every second?

Reply:
+1, agree, I'll see if it could monitor the change instead of doing the cal every second.
(Monitor kubernetes secret may require passing additional info such as secret name into webhook server)

Reply:
@przemeklal ,  Just added `fsnotify` to watch the file change, PTAL